### PR TITLE
simplify the reader for speedtest

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
simplify the reader for speedtest

## Motivation and Context
additionally, count only success operations,
truncated incomplete calls don't need to be
counted.

## How to test this PR?
Nothing should change just accuracy.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/peer-rest-server.go [hunk 1]
```diff
@@ -1137,30 +1137,14 @@ type SpeedtestResult struct {
...
-	}
-	if len(b) > bo.remaining {
-		b = b[:bo.remaining]
-	}
-	copy(b, bo.buf)
-	bo.remaining -= len(b)
-
-	atomic.AddUint64(bo.totalBytesWritten, uint64(len(b)))
+func (rr *randomReader) Read(b []byte) (int, error) {
+	copy(b, rr.buf)
```
Comment (klauspost) [inline cmd/peer-rest-server.go]: If len(rr.buf) < len(p) the upper part is not written to.

File: cmd/peer-rest-server.go [hunk 2]
```diff
@@ -1137,30 +1137,14 @@ type SpeedtestResult struct {
 	Error     string
 }
 
-// SpeedtestObject implements "random-read" object reader
-type SpeedtestObject struct {
-	buf               []byte
-	remaining         int
-	totalBytesWritten *uint64
+// randomReader returns random data upon every
```
Comment (klauspost) [inline cmd/peer-rest-server.go]: Not really.  It returns the same data on every read.

SSD compression may be able to pick that up.

Warp uses a small (random) circular buffer as input to an encryption algorithm. Quite fast and doesn't require a big in-mem buffer and large slow reads from "rand": https://github.com/minio/warp/blob/master/pkg/generator/scambler.go#L84

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/peer-rest-server.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/peer-rest-server.go]: Yes but that is allowed @klauspost i.e is it required that we should return exact bytes that was asked ?

I don't think that is necessary here.

File: cmd/peer-rest-server.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/peer-rest-server.go]: I am using a simpler implementation for now

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/peer-rest-server.go [hunk 3]
```diff
@@ -1137,31 +1137,8 @@ type SpeedtestResult struct {
...
-	if len(b) > bo.remaining {
-		b = b[:bo.remaining]
-	}
-	copy(b, bo.buf)
-	bo.remaining -= len(b)
-
-	atomic.AddUint64(bo.totalBytesWritten, uint64(len(b)))
-	return len(b), nil
+func newRandomReader(size int) io.Reader {
+	return io.LimitReader(rand.New(rand.NewSource(time.Now().UnixNano())), int64(size))
```
Comment (klauspost) [inline cmd/peer-rest-server.go]: Last time I checked this was very CPU intensive. Check this is a reasonable concurrency. If it hasn't changed the random function is quite slow and a clear bottleneck.

Comment (harshavardhana) [reply cmd/peer-rest-server.go]: okay @klauspost will check

Comment (klauspost) [reply cmd/peer-rest-server.go]: @harshavardhana Here is a random source, that is 10x the `math/rand` speed: https://github.com/minio/pkg/pull/20

**Comment by harshavardhana** [comment]
@krishnasrinivas PTAL

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**